### ISSUE TITLE
The docs link has been down for a long time: https://cdn.jsdelivr.net/gh/simplenlg/simplenlg@master/docs/javadoc/index.html

### ISSUE THREAD

**Issue by EricZinda** [simplenlg/simplenlg#91]
Here's what I get when I hit it:

~~~
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Frameset//EN" "http://www.w3.org/TR/html4/frameset.dtd">
<!-- NewPage -->
<html lang="en">
<head>
    <!-- Generated by javadoc (1.8.0_172) on Sun Jun 07 15:57:42 CEST 2020 -->
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <title>SimpleNLG 4.5.0 API</title>
<script type="text/javascript">
    tmpTargetPage = "" + window.location.search;
    if (tmpTargetPage != "" && tmpTargetPage != "undefined")
        tmpTargetPage = tmpTargetPage.substring(1);
    if (tmpTargetPage.indexOf(":") != -1 || (tmpTargetPage != "" && !validURL(tmpTargetPage)))
        tmpTargetPage = "undefined";
    targetPage = tmpTargetPage;
    function validURL(url) {
        try {
            url = decodeURIComponent(url);
        }
        catch (error) {
            return false;
        }
        var pos = url.indexOf(".html");
        if (pos == -1 || pos != url.length - 5)
            return false;
        var allowNumber = false;
        var allowSep = false;
        var seenDot = false;
        for (var i = 0; i < url.length - 5; i++) {
            var ch = url.charAt(i);
            if ('a' <= ch && ch <= 'z' ||
                    'A' <= ch && ch <= 'Z' ||
                    ch == '$' ||
                ch == '_' ||
                ch.charCodeAt(0) > 127) {
                allowNumber = true;
                allowSep = true;
            } else if ('0' <= ch && ch <= '9'
                    || ch == '-') {
                if (!allowNumber)
                     return false;
            } else if (ch == '/' || ch == '.') {
                if (!allowSep)
                    return false;
                allowNumber = false;
                allowSep = false;
                if (ch == '.')
                     seenDot = true;
                if (ch == '/' && seenDot)
                     return false;
            } else {
                return false;
            }
        }
        return true;
    }
    function loadFrames() {
        if (targetPage != "" && targetPage != "undefined")
             top.classFrame.location = top.targetPage;
    }
</script>
</head>
<frameset cols="20%,80%" title="Documentation frame" onload="top.loadFrames()">
<frameset rows="30%,70%" title="Left frames" onload="top.loadFrames()">
<frame src="overview-frame.html" name="packageListFrame" title="All Packages">
<frame src="allclasses-frame.html" name="packageFrame" title="All classes and interfaces (except non-static nested types)">
</frameset>
<frame src="overview-summary.html" name="classFrame" title="Package, class and interface descriptions" scrolling="yes">
<noframes>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<h2>Frame Alert</h2>
<p>This document is designed to be viewed using the frames feature. If you see this message, you are using a non-frame-capable web client. Link to <a href="overview-summary.html">Non-frame version</a>.</p>
</noframes>
</frameset>
</html>
~~~

**Comment by AayushSameerShah** [comment]
@EricZinda Hey, any update on this? I can't find any tutorial for this library.
Have you found any workaround to this?

Thanks!

**Comment by EricZinda** [comment]
No update that I've seen.  There's an [old google groups link](https://groups.google.com/g/simplenlg?pli=1) that still works and has a lot of helpful conversations on it, but appears to be abandoned now.  My last two questions weren't touched

**Comment by AayushSameerShah** [comment]
You are correct mate, and also NLG in Java is kind of dead.

**Comment by rogerdickey** [comment]
@AayushSameerShah @EricZinda what would you say is the best NLG system today? I understand LLMs can achieve similar functionality but I'm looking for something packaged in a similar way to SimpleNLG

**Comment by AayushSameerShah** [comment]
@rogerdickey Hey,
Firstly I want to clarify that I am not sure what this Simple NLG  library does. As said there is no clear documentation or thing that matter. So, I wonder that if even it is allowed to compare this library with HF transformers.

But if you are looking for the libraries that allows NLG and especially in Java, then there is "Spark NLP" by John Snow Labs. They have a large collection of supported models like T5 and GPT2, tuned or raw for specific usecases. But they don't have rich support of models like Hugging Face does.

If you have access to python then there are plenty ranging from Hugging Face Transformers (of course) and the most recent works done in the GGUF related models ie. LLaMa CPP.

Well, I should also briefly touch on this. LLaMa CPP mainly written in CPP but it has its wrappers inany languages - including JAVA and PYTHON.

Now, it support many models. From LLAMA2 to Mistral, Zypher, Falcon and a lot more. Also it supports CPU inferencing. So, even if you don't have GPU, on CPU it can work pretty fast.

Let me know if I can further help.

**Comment by rogerdickey** [comment]
Thanks for your reply @AayushSameerShah . I'm actually looking for a Node JS package or library that will generate english in a procedural way (eg take a list of lemmas and "realize" a coherent sentence with correct conjugations, tenses, plurals, articles, etc) rather than using an LLM. LLMs are too general purpose and expensive to run and all I'm doing is generating grammatically correct sentences and want to provide a free experience to the user. I checked out Spark and it appears to have NLU capabilities but nothing for NLG. I found SimpleNLG and jsRealB that appear to do what I need but neither is currently supported actively. Please lmk if you're aware of a non-AI-powered NLG system that is actively supported and can do the above!

Separately, I'm also looking for an NLU system that can take a sentence and parse it into a dependency tree by POS. There are a number of older systems that don't appear to have been touched in a while (like OpenCCG), there's Nodebox which is Python based and has limited first-order "chunking" capabilities, and there's Spark which appears well supported and claims to have this functionality but I can't find any examples online.

Any guidance appreciated!

**Comment by rdeoliveira** [comment]
@rogerdickey,  seems realisers like SimpleNLG are just what you need for the NLG bit of your application. LLMs could cause all sorts of issues such as hallucinations, generate not enough or too much text, and/or incur in high bills as you pointed out.

You are also right that SimpleNLG hasn't been maintained in a while; I personally haven't pushed a commit in ages. But it was pretty robust when I used to use it, so I don't see a reason to not use it as is. Regarding documentation (which seems to be what this thread was originally about), have a look at https://github.com/simplenlg/simplenlg/wiki/Section-III-–-Getting-started as well as the many unit tests at https://github.com/simplenlg/simplenlg/tree/master/src/test/java/simplenlg/syntax/english. You mentioned you're looking for a Node JS implementation; I'm unfortunately unaware of a good realizer in JS, but does seem like jsRealB is a good candidate even though development is not active. Have you tried it?

For the NLU bit, again I'm unaware of any good parsers for Node JS but can certainly recommend the Python-based https://spacy.io. There seems to be a Node JS wrapper https://www.npmjs.com/package/spacy-nlp but I've never used it myself.

Good luck!

**Comment by EricZinda** [comment]
@AayushSameerShah, my realization needs were very light so I ended up going with https://github.com/jaraco/inflect which mostly helps you just figure out the right form of different words.

However, one realization resource I looked at that is *very* robust, has been around forever, and still under active development (but also has a steep learning curve) is: [KPML](http://www.fb10.uni-bremen.de/anglistik/langpro/kpml/README.html)

**Comment by rogerdickey** [comment]
@rdeoliveira @EricZinda I haven't tried jsRealB yet but I'm going to check that out and fall back to SimpleNLG if it doesn't work. Maybe I'll publish my JS wrapper. KPML looks awesome (love the KR languages) but could be too complex for my use case. Will also try Spacy!

**Comment by AayushSameerShah** [comment]
@rogerdickey Hey,

And probably NLTK too! 🤗

**Comment by Saad-Mahamood** [comment]
Hi all. Sorry for not responding earlier. I will take a look at fixing the links. The raw HTML files are still there in the repo and can be accessed if the repo is cloned.

**Comment by Saad-Mahamood** [comment]
Now fixed in PR #93.
